spectral_cube/dask_spectral_cube.py

```diff
@@ -422,24 +486,41 @@ def apply_function_parallel_spectral(self,
             representing the mask.  It may also accept ``**kwargs``.  The
             function must return an object with the same shape as the input
             spectrum.
+        accepts_chunks : bool
+            Whether the function can take chunks with shape (ns, ny, nx) where
+            ``ns`` is the number of spectral channels in the cube and ``nx``
+            and ``ny`` may be greater than one.
+        save_to_tmp_dir : bool
+            If `True`, the computation will be carried out straight away and
+            saved to a temporary directory. This can improve performance,
+            especially if carrying out several operations sequentially. If
+            `False`, the computation is only carried out when accessing
+            specific parts of the data or writing to disk.
         kwargs : dict
             Passed to ``function``
         """
 
-        data = self._get_filled_data(fill=self._fill_value)
-
-        # apply_along_axis returns an array with a single chunk, but we need to
-        # rechunk here to avoid issues when writing out the data even if it
-        # results in a poorer performance.
-        data = data.rechunk((-1, 50, 50))
-
-        newdata = da.apply_along_axis(function, 0, data, shape=(self.shape[0],))
+        def wrapper(data, **kwargs):
+            if data.size > 0:
+                return function(data, **kwargs)
+            else:
+                return data
 
-        return self._new_cube_with(data=newdata,
-                                   wcs=self.wcs,
-                                   mask=self.mask,
-                                   meta=self.meta,
-                                   fill_value=self.fill_value)
+        if accepts_chunks:
+            return self._map_blocks_to_cube(wrapper,
+                                            rechunk=(-1, 'auto', 'auto'), **kwargs)
+        else:
+            data = self._get_filled_data(fill=self._fill_value)
+            # apply_along_axis returns an array with a single chunk, but we
+            # need to rechunk here to avoid issues when writing out the data
+            # even if it results in a poorer performance.
+            data = data.rechunk((-1, 50, 50))
```

Comment:
Is (50, 50) just a reasonable compromise for the write out? I'm wondering if there is any benefit to being able to change the spatial dim sizes.

Reply:
It was a reasonable compromise for the datasets I tried, but I think I could perhaps write it more generically, will give it a shot.

Reply:
See #632